ee/elf-loader/src/loader/loader.c

```diff
@@ -0,0 +1,73 @@
+/*
+# _____     ___ ____     ___ ____
+#  ____|   |    ____|   |        | |____|
+# |     ___|   |____ ___|    ____| |    \    PS2DEV Open Source Project.
+#-----------------------------------------------------------------------
+# (c) 20020 Francisco Javier Trujillo Mata <fjtrujy@gmail.com>
+# Licenced under Academic Free License version 2.0
+# Review ps2sdk README & LICENSE files for further details.
+*/
+
+#include <kernel.h>
+#include <loadfile.h>
+#include <sifrpc.h>
+#include <errno.h>
+
+//--------------------------------------------------------------
+//Start of function code:
+//--------------------------------------------------------------
+// Clear user memory
+// PS2Link (C) 2003 Tord Lindstrom (pukko@home.se)
+//         (C) 2003 adresd (adresd_ps2dev@yahoo.com)
+//--------------------------------------------------------------
+static void wipeUserMem(void)
+{
+	int i;
+	for (i = 0x100000; i < GetMemorySize(); i += 64) {
+		asm volatile(
+		    "\tsq $0, 0(%0) \n"
+		    "\tsq $0, 16(%0) \n"
+		    "\tsq $0, 32(%0) \n"
+		    "\tsq $0, 48(%0) \n" ::"r"(i));
+	}
+}
+
+//--------------------------------------------------------------
+//End of func:  void wipeUserMem(void)
+//--------------------------------------------------------------
+// *** MAIN ***
+//--------------------------------------------------------------
+int main(int argc, char *argv[])
+{
+	static t_ExecData elfdata;
+	int ret;
+
+	if (argc < 1) {  // arg1=path to ELF
+		return -EINVAL;
+	}
+
+	// Initialize
+	SifInitRpc(0);
+	wipeUserMem();
+
+	//Writeback data cache before loading ELF.
+	FlushCache(0);
+	ret = SifLoadElf(argv[0], &elfdata);
```

Comment:
This is the code for loading the APP2 elf. Using a function instead of custom code. Also note the special wipeUserMem fuction, shouldn't the sections already be wiped when loading the ELF, as the cusom loading code above does?

Reply:
Not sure if I understood your question here...
I think that `SIfLoadElf` just wipe the sections where `ELF` is going to be loaded, but I suppose that is safer if the whole user-space memory is clean before executing the APP2

Reply:
Sony also does this memory wipe. It's good for ensuring consistent behaviour. krHACKen observed that SifLoadElf doesn't zero the unloaded sections.
This memory wipe would also be the one to erase the memory used by the program's stack.